### DISCUSSION TITLE
How to mix Client & Server Side Auth

### DISCUSSION THREAD

**Discussion by hmajid2301** [pocketbase/pocketbase#2153] [Q&A] [answered]
Hi,

First of all can I say how great PocketBase is I'm a big fan!
I'm building an application using SvelteKit and PocketBase and I have a few question(s).

Sorry if any of these seem really obvious, but I feel like I'm missing something.

So I know SvelteKit allows us to use SSR, and it makes a lot of sense with the `hooks.server.ts`  file and
also in the `load` function for some pages. As we can fetch the data server-side rather than client-side.

Even with form actions it does make a bit of sense, that this is handled server side after a bit of validation.
We make a request to pocketbase.

However for some actions such as user clicking a button to delete an item. I am currently
making a fetch request -> `+server.ts` -> `my pocketbase backend
Part of the reason I've built it like this is because all the auth related logic is on the server-side.
So to auth properly with pocketbase at the moment it needs to do this.

I was wondering for these actions would it make more sense for the client to make a direct request
to the pocketbase server. If this is the case, how can I auth on the client side? I cannot access the cookies
atm in JS. I believe this is due to them being `httponly`. 

In my specific case, I deploy my site to Netlify and my PocketBase app. So it feels like there is wasted
network request

Browser -> Netlify Functions -> Fly (PocketBase)

(I am imaging if the DB were external to PocketBase, then there'd be yet another network request).

**tl;dr: It seems the way I have built my application at the moment, is making a redundant network request.**

- Does it make sense to send requests to pocketbase from the client
- If so how can we authenticate the client and also on the server side
- When does it make sense to use SSR
  - atm I'm thinking load functions and form actions_

Sorry if some of these might be more SvelteKit specific questions rather than strictly PocketBase. I'm just a bit confused. Any help would be greatly appreciated. If some of these questions don't make sense, please let me know.

My Hooks: https://gitlab.com/bookmarkey/gui/-/blob/main/src/hooks.server.ts

Example Load: https://gitlab.com/bookmarkey/gui/-/blob/main/src/routes/my/collections/%5Bid%5D/%2Bpage.server.ts#L27-36 (getting data from PocketBase, using the js-SDK)

**Comment by ganigeorgiev** [comment, marked as answer]
You can have a "mixed" (both browser and node) access to the PocketBase server.

If you want to maintain a shared auth state in both browser and node, you'll have to export the cookie with `httpOnly: false` and have 2 SDK instances. The browser SDK instance can be a global singleton since the browser client is only the user, something like:

```js
import PocketBase from "pocketbase";

const pb = new PocketBase("http://127.0.0.1:8090");

pb.authStore.loadFromCookie(document.cookie)

pb.authStore.onChange(() => {
  document.cookie = pb.authStore.exportToCookie({ httpOnly: false }); // make sure to export with httpOnly: false also on the node client
});
```

Whether it make sense - you can check https://github.com/pocketbase/pocketbase/discussions/2046 for some guidance.

I don't feel confident giving a definite/hard answer as it really depends on your use case and what you prefer working with.

It is OK to have a "mixed" (browser and node) approach, especially in case you want to use the realtime service, because otherwise you'll have to create a node proxy for the realtime connection.
It makes sense to have a node middleware service and/or use SSR, if you don't like or feel comfortable working with Go and need to implement custom business logic, eg. you want to perform some extra form data validations, request transformations, handle payment webhooks, etc. (_note that extending PocketBase with js app event hooks is in the [roadmap](https://github.com/orgs/pocketbase/projects/2)_).

I personally prefer working with SPA and static/prerendered htmls and avoid SSR whenever I can because for me it is just easier and there is no need to maintain another server application _(Netlify/Vercel helps with that to some extend, but I usually deploy on my own VPS)_.

**Reply by hmajid2301** [reply]
In terms of writing Go, I have no issue with that! In fact I actually started this personal project to learn Go. But PocketBase has been so good, that I have barely written any. 

Yeh thinking about it some of those questions are better aimed at the SvelteKit community. 

Thank you for the quick reply!
